### ISSUE TITLE
Sequentially running systems on other threads?

### ISSUE THREAD

**Issue by Ashe** [Doraku/DefaultEcs#74]
Hey there. So I usually try to get the rendering of my games on another thread, as this process tends to not do any kind of updating or `Set`ting of components. However, the examples of the usage of a `runner` have only shown singular `System`s.

I have a camera, and I'm thinking of making a `CameraSystem` which takes the `Entity` with the component `Camera`. This will then call the `SpriteBatch.Begin` process using the camera's properties to view the game world.

After that, I would then sequence all my other world-space render calls so that they happen after the `Begin` call and before the `End` call. This would literally be the same as I would do it on my main thread anyway.

I'm not very clear on how `Runner`s really work. I get that it allows you to run a system on another thread with ease, but I still need to call `Update` otherwise my game stops rendering, so when exactly does the system `Update`? If it waits for my call to it from my main thread, then surely it isn't multithreaded? Little bit confused and would appreciate some help.

1. What happens if you use the same runner for multiple systems? Can you sequence them?
2. If not, would there be a way to use standard C# multithreading techniques to sequence the running of systems (that literally get and render, and don't interfere with logic at all?)
3. What exactly does `Update` do when a system is associated with a runner? Why is it necessary to call it?

Thanks very much for your hard work!

**Comment by Doraku** [comment]
So I will have to answer your questions in the wrong order ^^

> 3\. What exactly does `Update` do when a system is associated with a runner? Why is it necessary to call it?

The main goal of `SystemRunner` is to split the process of a System on multiple thread so it is processed faster, not to run a system in the background. To give you a concrete example with an `AEntitySystem`:
```
[With(typeof(Velocity), typeof(Position)]
public sealed class VelocitySystem : AEntitySystem<float>
{
    public VelocitySystem(World world, SystemRunner<float> runner)
        : base(world, runner)
    { }

    protected override void Update(float elaspedTime, in Entity entity)
    {
        ref Velocity velocity = ref entity.Get<Velocity>();
        ref Position position = ref entity.Get<Position>();

        Vector2 offset = velocity.Value * elaspedTime;

        position.Value.X += offset.X;
        position.Value.Y += offset.Y;
    }
}
```
This is a pretty basic system to update a position with a velocity. When you call its main Update method, it will run the `Update(float elaspedTime, in Entity entity)` on each entity of its internal `EntitySet`. With a `SystemRunner`, this `EntitySet` will be equally split on its threads, speeding up this process.
The next interesting system class using `SystemRunner` is the [ParallelSystem](https://github.com/Doraku/DefaultEcs/blob/master/documentation/api/DefaultEcs-System-ParallelSystem-T-.md), it takes a collection of systems that will be run in parallel on the threads of the `SystemRunner`, making sure that once you call its `Update` method, all the inner systems will also be updated once.

> 1. What happens if you use the same runner for multiple systems? Can you sequence them?

There is two main system types to help you build a workflow of the update of your game:
- [SequentialSystem](https://github.com/Doraku/DefaultEcs/blob/master/documentation/api/DefaultEcs-System-SequentialSystem-T-.md): define systems which need to be run one after the other because they are dependent of the execution of one another
```
// example in DefaultSlap sample
            _runner = new SystemRunner<float>(Environment.ProcessorCount);
            _system = new SequentialSystem<float>(
                new PlayerSystem(Window, _world), // handle the player input
                new HitSystem(_world), // check collisions to kill enemies, need the position and state of the player
                new GameSystem(_world), // spawn new enemies
                new AISystem(_world, _runner), // move the enemies
                new PositionSystem(_world, _runner), // just translate the position to screen coordinates
                new DrawSystem(_batch, _square, _world)); // draw everything
```
- [ParallelSystem](https://github.com/Doraku/DefaultEcs/blob/master/documentation/api/DefaultEcs-System-ParallelSystem-T-.md): systems which can be run in parallel of each other (there is a constructor which take a system before the `SystemRunner`, this is in case you need a system to run on the main thread (like operations on the spriteBatch for example)
```
// example of what you want to achieve, but I am not sure of the SpriteBatch is thread safe, at least the End method needs to be called on the main thread I think
            _runner = new SystemRunner<float>(Environment.ProcessorCount);
            _system = new SequentialSystem<float>(
                new ActionSystem(_ => spriteBatch.Begin()), // ensure Begin is called before the other systems run
                new ParallelSystem(_runner,
                    new IndependantSystem1(),
                    new IndependantSystem2(),
                    new IndependantSystem3()),
                new ActionSystem(_ => spriteBatch.End()));
```
You simply control the flow of your game by calling `Update` on the main system. Refer [here](https://github.com/Doraku/DefaultEcs#systemrunner) for where it is safe to use a `SystemRunner` (or tell me if it need more explanation).

There is nothing wrong with reusing the same `SystemRunner` instance for multiple systems (it's actually better to do so), as long as those systems don't run at the same time!
```
// bad, don't mix it, you would probably end up in a deadlock
new ParallelSystem(_runner,
    new PositionSystem(_world, _runner),
    new AISystem(_world, _runner));
```

> 2\. If not, would there be a way to use standard C# multithreading techniques to sequence the running of systems (that literally get and render, and don't interfere with logic at all?)

I believe that when you want to render your game state on the screen, you want what appear to be determined, as in not dependent on whether a logic system has finished running or not, you want your state to be stable. You could use a task to run a system in the background
```
Task.Run(() =>
{
    while (true)
    {
        system.Update(whatever);
    }
});
```
but you will probably not like what you will end up on screen (or this system need to be completely independent of everything else). Using SequentialSystem/ParallelSystem to define your workflow help you to synchronize systems between each others so that for each update or frame, your game has run exactly one step, and everything is stable. You could define your own types to create a workflow if you do not want to use mine but I suspect you would end up with something really close.

Hopefully I didn't forget anything here :)

**Comment by Ashe** [comment]
As always, thank you for your explanation. I didn't learn that the `SystemRunner` was to make a single system run on multiple threads, I just thought it was to run the single system on it's own thread.

Knowing this I now understand the significance of the sequential system. From the readme where it talks about how runners can't be used with the sequential system, I got the impression that it wasn't multithreaded and that I shouldn't use it. While it's semi true, I now know that you use multithreading to speed up systems inside the sequence and not the sequence itself.

I wonder if using the sequence with runners will be what I want then. Does one system fully end before the next one begins when sequenced and using runners? This does explain why `Update`has to be called each frame, when normally with multithreading you just tell the thread to 'go' and let it do it's thing.

I'll play around a little more but thanks for the answers!

**Comment by Doraku** [comment]
In a SequencialSystem, each inner system need to finish its update before the next one can start its own.

In an engine like unity, I think you just declare your system dependencies (like the collision system need to run after the move system ect...) and the update workflow is automatically built for you based on those rules and component dependencies (for example it would not try to run in parallel a system that read a component with an other system which write this component). Personally I prefer to explicitely define my workflow which while it might be harder to setup give you better control on how things run (that's my opinion anyway but I am just a hobbyist :p).

Keep in mind that even though you can run system in parallel with ParallelSystem, it will probably have a worse repartition on your threads than trying to parallelise each system individually. This is because for AEntitySystem, each thread will have the same amount of work so they should end their processing relatively close to each other while with the ParallelSystem, two of its inner systems can have completely different process time, making your cpu iddle while the last system finish its update (I have plan to use this iddle time for schedulable jobs but I still need to work on it).

**Comment by Ashe** [comment]
Thank you again for your help! So far I keep my structs immutable so I `Set` them every time I want to overwrite them but I guess if I need to use the system runners I really need to not do that. Why is there a requirement with using `Set` with system runners anyway, is it because it could change an `EntitySet` if a new `Component` gets added to an `Entity`?

**Comment by Doraku** [comment]
Yes, when you `Set` a component value, an event will be sent to interested EntitySets so they can check the modified Entity (either Added or Changed). EntitySet and the class handling the component storage are not thread safe because I am lazy (and putting lock everywhere can't be good for performance) so you don't want those to happen in parallel.
Note that even if your components are immutable you can still use the quick and dirty `Get` to change their value
```
// Get returns a ref Component, you can see it as a Component*, this is its real address, not a copy
entity.Get<Component>() = new Component();
```
The Get does not do any check, do not send any event, and each entity should work on its own memory location so parallel processing is ok.
If you want to use parallel processing but need to make composition changes on entities, you can use the [EntityCommandRecorder](https://github.com/Doraku/DefaultEcs#Overview_Command) to do so.

**Comment by Ashe** [comment]
This is where my C# knowledge isn't so great, I find it difficult to see what's passed by reference or whatever. Is the code you posted the same as this (forgive me if it's not correct I'm not at my pc)
```cs
var ref foo = ref entity.Get<Component();
<do things with foo and generate bar>
foo = bar;
```
If `foo` is a reference to the component from `Get`, does saying `foo = bar` change the component itself as per your example, or change the reference `foo` to reference `bar` instead? I'll get used to it but in C++ I feel a lot more secure with what type something actually is and if you hadn't have said that then I wouldn't have known I could overwrite an immutable component without using `Set`, you're a wealth a knowledge so thanks again!

**Comment by MarkWilds** [comment]
> entity.Get<Component>() = new Component();
This feels weird, does it really work?
To my understanding it returns a ref local value type, the ref is assigned a new value but nothing is done with it.

**Comment by Doraku** [comment]
It uses [ref return](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/ref-returns) which is a pretty newish (C#7 august 2017) and really specific in its usage so it is probably not well known. DefaultEcs was born early 2018 as I wanted to play around with this feature, my game prototype used the more classic object approach but I was reading a lot on the ECS pattern and I thought it was a match made in heaven to reduce copy and use struct for better cache locality and so it was born (and my game development slowed down to a crawl ^^")

So yes the `Get<Component>` gives you a `ref Component` to the internal array of storage of the actual value of the component for a given entity, really handy for struct as they are normally returned by copy (technically you could argue that even class types are returned by copy, a copy of the reference to the actual object that is) so you can't directly modify its value but with ref return, you can :)

@Ashe it is `ref var foo = ref entity.Get<Component>()` but yes, foo is a reference to the internal array of DefaultEcs, the real value of Component for entity so if you assign a new Component value to foo, you are actually changing the value for entity and don't need to use `entity.Set(bar)`

**Comment by Doraku** [comment]
Closing this. Btw if you have small questions on the library I created a [gitter chat](https://gitter.im/Doraku/DefaultEcs) if it does not warrant a new issue (not that I would mind anyway ^^).
